goasupport/generator.go

```diff
@@ -0,0 +1,387 @@
+		return nil, err
+	}
+
+	if err := ctxWr.ExecuteTemplate("generateETag", generateETag, nil, requestContexts); err != nil {
+		return nil, err
+	}
+
+	for _, responseCtx := range responseContexts {
+		if err := ctxWr.ExecuteTemplate("conditional", conditional, nil, responseCtx); err != nil {
+			return nil, err
+		}
+		if err := ctxWr.ExecuteTemplate("modifiedSince", modifiedSince, nil, responseCtx); err != nil {
+			return nil, err
+		}
+		if err := ctxWr.ExecuteTemplate("matchesETag", matchesETag, nil, responseCtx); err != nil {
+			return nil, err
+		}
+		if err := ctxWr.ExecuteTemplate("setLastModified", setLastModified, nil, responseCtx); err != nil {
+			return nil, err
+		}
+		if err := ctxWr.ExecuteTemplate("setETag", setETag, nil, responseCtx); err != nil {
+			return nil, err
+		}
+		if err := ctxWr.ExecuteTemplate("setCacheControl", setCacheControl, nil, responseCtx); err != nil {
+			return nil, err
+		}
+	}
+	for _, entity := range entities {
+		if err := ctxWr.ExecuteTemplate("getLastModified", getLastModified, nil, entity); err != nil {
+			return nil, err
+		}
+		if err := ctxWr.ExecuteTemplate("getETagData", getETagData, nil, entity); err != nil {
+			return nil, err
+		}
+	}
+	err = ctxWr.FormatCode()
+	if err != nil {
+		return nil, err
+	}
+	return []string{ctxFile}, nil
+}
+
+const (
+	constants = `
+	const (
+	// IfModifiedSince the "If-Modified-Since" HTTP request header name
+	IfModifiedSince = "If-Modified-Since"
+	// LastModified the "Last-Modified" HTTP response header name
+	LastModified = "Last-Modified"
+	// IfNoneMatch the "If-None-Match" HTTP request header name
+	IfNoneMatch = "If-None-Match"
+	// ETag the "ETag" HTTP response header name
+	// should be ETag but GOA will convert it to "Etag" when setting the header.
+	// Plus, RFC 2616 specifies that header names are case insensitive:
+	// https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+	ETag = "Etag"
+	// CacheControl the "Cache-Control" HTTP response header name
+	CacheControl = "Cache-Control"
+	// MaxAge the "max-age" HTTP response header value
+	MaxAge = "max-age"
+)`
+	conditionalResponseEntity = `
+	// ConditionalResponseEntity interface with methods for the response entities
+type ConditionalResponseEntity interface {
+	// returns the time of last update 
+	GetLastModified() time.Time
+	// returns the values to use to generate the ETag
+	GetETagData() []interface{}
+}`
+
+	cacheControlConfig = `
+   type CacheControlConfig func() string 
+   `
+	conditional = `
+{{ $resp := . }}
+// Conditional checks if the entity to return changed since the client's last call and returns a "304 Not Modified" response
+// or calls the 'nonConditionalCallback' function to carry on.
+func (ctx *{{$resp.Name}}) Conditional(entity ConditionalResponseEntity, cacheControlConfig CacheControlConfig, nonConditionalCallback func() error) error {
+	lastModified := entity.GetLastModified()
```

Comment:
Technically a lot of the logic in these Context specific methods could be moved out into a common function that is generated instead of repeating it. Doesn't really matter since noone will change it, but I think go/goa has a bit of an idium, even tho the code is generated, it shouldn't be shit. :)

e.g. the main logic in this method could be moved to a non exposed function doConditional(ctx, entity, cacheControl, nonConditional). similar ctx.matchesTag could be matchesTag(ctx, tag)